### ISSUE TITLE
More ergonomic API for printing formatted text

### ISSUE THREAD

**Issue by Cryptjar** [murarth/mortal#7]
# Motivation

While Mortal has a rich feature set for printing formatted text to the terminal, the API is somewhat cumbersome to use. Especially, when comparing the API of Mortal with other terminal-interface crates, it seems to be the most complicated choice. One striking feature of other crates is the existence of style modifiers that implement the `Display` trait, which allows to simply include them in the `println!` and other macros (subsequently called the `Display` strategy).

For example:
```rust
// The Display strategy (Termion style)
println!("{r}Red{reset} {bold}{g}Green{reset} {b}{}{reset}", 42,
    r=color::Fg(color::Red), g=color::Fg(color::Green), b=color::Fg(color::Blue),
    bold=style::Bold, reset=style::Reset);
```
Outputs a red `Red`, a green and bold `Green`, and a blue `42`.

The Mortal equivalent would be something like:
```rust
// Current Mortal style
terminal.write_styled(Some(Color::Red), None, Style::default(), "Red");
terminal.write_str(" ");
terminal.write_styled(Some(Color::Green), None, Style::BOLD, "Green");
terminal.write_str(" ");
terminal.write_styled(Some(Color::Blue), None, Style::default(),
    &format!("{}", 42));
```

In comparison Mortal needs quite a lot code to achieve the same output. It become especially messy when writing text where a multiple words need to be formatted differently than the surrounding. The Mortal code is also somewhat harder to read and comprehend.

# Problem

While implementing the `Display` strategy has a positive effect on usability, it has also drawbacks. First, it allows to write those formatting tokens to other `Write`-rs such as files, which could lead to unintended outcomes. Second, this approach works especially when encoding the formatting directly into the output text as it is with the ANSI escape sequences, which also could cause issues when specific escape sequences aren't supported (which in turn is harder to test due to the decoupeling of formatting and terminal outputting).

```rust
// Writing formatting into a file (Termion style)
let mut file = File::create("foo.txt").unwrap();
writeln!(file, "{r}Red{reset} {bold}{g}Green{reset} {b}{}{reset}", 42,
    r=color::Fg(color::Red), g=color::Fg(color::Green), b=color::Fg(color::Blue),
    bold=style::Bold, reset=style::Reset);
// Yields: '[38;5;1mRed[m [1m[38;5;2mGreen[m [38;5;4m42[m'
```

However, Mortal has an approach which does not depend on ANSI escape sequences and utilizes syscalls where appropriated. The latter would especially cause unintended behavior when implementing the `Display` strategy in combination with concurrency and non-terminal outputs. Therefore, the `Display` strategy is not suitable for Mortal.

In order to improve the usability of Mortal without changing the concepts of Mortal and without the `Display` strategy, a more explicit function-calling approach is needed. Considering the examples above, Mortal needs the opportunity to switch the color of the terminal (potentially via a syscall) while writing the complex string to the terminal. Hence, the straightest way would be to split the text at formatting boundaries and issue different API calls for each string fragment. This would consequently lead to something similar to the API as it is.

Nevertheless, there are options to improve the API. First, the amount of code and unnecessary repetitions could be reduced. Second, it would be nice to have a macro to further simplify formatting. Obviously, `println` and others are not a good option, thus a custom macro would be needed.

# Proposed Solution

In order to reduce repetition regarding format definitions, a new type capsuling `Style` and foreground and background `Color`s can help when the same format is used in several places and can be used to refer to a format as a whole. Such a type might be as simple as:

```rust
pub struct Theme {
    pub style: Style,
    pub fg: Option<Color>,
    pub bg: Option<Color>,
}
```

Yet, that new type can simplify the above example:
```rust
let red = Theme::color(Color::Red);
let green = Theme::color(Color::Green).style(Style::BOLD);
let blue = Theme::color(Color::Blue);
// Using themes
terminal.write_themed(&red, "Red");
terminal.write_str(" ");
terminal.write_themed(&green, "Green");
terminal.write_str(" ");
terminal.write_themed(&blue, &format!("{}", 42));
```

A further improvement could bring a builder-pattern like printer which would essentially wrap the `TerminalWriteGuard` (if not directly implemented there):
```rust
let red = Theme::color(Color::Red);
let green = Theme::color(Color::Green).style(Style::BOLD);
let blue = Theme::color(Color::Blue);
let printer = Printer::lock(&terminal).unwrap();
// Using a builder-patter like printer
printer.themed(&red, "Red").str(" ").themed(&green, "Green").str(" ")
    .themed(&blue, &format!("{}", 42));
```

Finally, adding a custom macro (using `macro_rules`) could minimize coding. For the formatting of variables (such as the blue 42 case) such a macro could easily include the `format!` part. The syntax for such a macro might be a optional `Theme` variable followed by a format string and optionally format arguments:
```
// pseudo code proposed syntax for using the themed macro:
themed!(terminal;
  [theme]:"format_str"[, format_arg[, ...]];
  [...]
);
```

This macro would reduce the formatting of our example down to a single macro invocation:
```rust
let red = Theme::color(Color::Red);
let green = Theme::color(Color::Green).style(Style::BOLD);
let blue = Theme::color(Color::Blue);
let printer = Printer::lock(&terminal).unwrap();
// Using a custom macro
themed(printer;
    red:"Red"; :" "; green:"Green"; :" "; blue:"{}", 42;
);
```
That would expand for instance to the same as the previous example.

Ignoring the theme definitions (since they are reusable), the last example using a custom macro beats even the introducing example in length of code.

_The proposed concepts were prototypically implemented and tested for feasibility._

# Unresolved Questions

Open and further options:
- Whether to put the additional types and the macro in to a separate utilty crate. This would keep the primary crate cleaner and users might easier opt-out that feature, but it causes more maintenance work.
- The exact syntax for the proposed macro has room for alternatives and even further short cuts.
- Should be there multiple macros. For instance one for using `Terminal` and one for `Printer`/`TerminalWriteGuard`. Such a differentiation might also be implemented within the same macro e.g. via prefixing the terminal/printer variable.
- Should the proposed builder-pattern printer be implemented directly in `TerminalWriteGuard`.
- There are also alternative concepts, for instance Crossterm uses special structs (`StyledObject`) that capsule the text to be formatted with the format definition for it. These struct themselves implement again `Display`, but that approach might also be used similar to the here proposed concept, since those structs already capsule text with the same format and make the format boundaries explicit.
- How and where to apply proper error handling (whether to propagate a `Result`, or just `unwrap` intermediate results). On the one hand, the macro (similar to `println`) might rather unwrap than returning a `Result`. On the other hand, the builder-pattern printer might rather return `Result` but needs by design to return itself, and should not introduce additional coding overhead. Thus, the builder-pattern printer needs a compromise, maybe something like final `finish` or `flush` call that returns a `Result` for all previous cascading calls.
- Whether implementing the builder-pattern printer as a preparing cascade which needs a final execution call, or as an executing-when-called cascade. While the former might be more difficult to implement and maybe to use (the final call must not be omitted), it would ease the question for error handling, and potentially allow to reuse such prepared states for printing similar or identical text passages.

**Comment by murarth** [comment]
I agree that the existing API for styled text output is cumbersome. As you've noted, the underlying implementation of terminal attributes makes it difficult to output styled text using the `std::fmt` API.

The proposed solutions are somewhat an improvement, reducing some of the noise of repeated function names and `None` parameters required in the existing API.

However, I think, the ideal solution would be a procedural macro that parses a format string, akin to the standard `format_args!` macro. Use of such a macro could look something like this:

```rust
// Would generate code to set style attributes and output text;
// delegates to std::fmt for displayed values.
term_writeln!(terminal, "<red>Red<reset> <green bold>Green<reset> <blue>{}<reset>", 42);
// Can use color and style values from input parameters
term_writeln!(terminal, "<fg={} bg={}>Foo<reset>", Color::Red, Color::Black);
term_writeln!(terminal, "<style={}>Bar<reset>", Style::BOLD);
```

Unfortunately, procedural macros in an expression context are currently unusable or unstable (I'm not sure which).

**Comment by Cryptjar** [comment]
## tl;dr

Unfortunately, I have to contradict you. I think your counterproposal of putting formatting syntax directly into the format string is rather the wrong way instead of something like the ideal solution. Nevertheless, I have to admit, that in simple cases your proposal simplifies things and even increases readability of code (for the crate users).

But on the other hand, in my opinion, your proposal ignores the capabilities provided by the Rust compiler and requires Mortal to build its own parser. Additionally, I claim that in the long run your proposal not only makes it difficult to implement that feature in the Mortal crate but also handicaps the usability of the crate for the user.


## Scope

Short clarification on terms, there are Rust `std::fmt` format strings, which are used to inject variables into the output string, lets just call them Rust Format Strings (RFS).
And there are (future) Mortal formatting tokens, which are intended to specify the color and style of terminal output, lets just call them Mortal Formatting Tokens (MFT).

In my subjective point of view, I was always and still are proposing to separate the MFTs form the RFSs, by splitting the RFSs at the style/color boundaries and put into those gaps the MFTs. This way the RFSs are string literals (as they always are) and the MFTs are actual syntactic tokens -- Rust compiler tokens. Your proposal opposes this idea by essentially putting the MFTs into the same string literals as the RFSs, resulting in one single big string literal containing both.

Following, I will try to elaborate why it might not be a good idea to mix the MFTs and RFSs within the same string literals, and why I believe that designing the MFTs as actual token, in the scope of the Rust compiler, has lots of advantages. I'll try to be neutral in terms of syntax, but make a hard differentiation between string literals and other tokens.

As a side note, I hope it is established that evaluation of the MFTs shall happen at compile time and not at run time. In that mean that an MFT is translated by the Rust compiler / macro / what-ever to an API-call that will apply the specified style/color to the terminal.



## String Literals

Your proposal obviously puts the MFTs into the RFS as stringified tokens. This design decisions has specific consequences. Most importantly, it requires parsing of the unified string literal. Since a string literal is only one single token to the Rust compiler, and since the `macro_rules!`-based macros work on those tokens only, it is not possible to parse the string literal with a `macro_rules!`-based macro, but instead needs procedural-or-what-ever macro, as you already noted.

But that is not the end, it is just the beginning of difficulties. The parsing part is the real mammoth task. First, one has to (well) define a syntax for the stringified MFTs. That definition should be simple enough to be usable, but also completed by allowing to escape those sequences for instance to allow
printing actual HTML code to the terminal.

Next, the, say, procedural macro has to implement a parser being capable of parsing that definition, and transform the input literal accordingly. That as I mentioned will end in splitting the input literal into several sub-strings, and put the terminal color/style changing API calls between printing those sub-strings.

Just consider one of the `std::fmt` examples:

```rust
format!("{1} {} {0} {}", 1, 2); // => "2 1 1 2"
```

Now, imagine that we introduce some simple MFTs:

```rust
term_write!(terminal, "{1} <red>{} {0}<reset> {}", 1, 2); // => "2 1 1 2"
```

The procedural macro now would need to split the string into three parts, and issue a `terminal.set_fg(Some(Color::Red))` and `terminal.clear_attributes()` respectively between the sub-strings. The difficult part is to match to the sub-strings `"{1} "`, `"{} {0}"`, `" {}"` to the correct arguments for the `std::fmt` functions. Essentially, it also requires to parse the RFS as Rust would do it anyway, but only in order to figure which argument(s) belongs to which sub-string, and rearranging them  where necessary (e.g. `"{1} "` => `"{} "`).

At this point we need already two parsers, and while the syntax for the MFT can be arbitrarily chosen, the other is tied to the official RFS and supposedly any user would expect to use all the RFS features even in the context of Mortal (and for a good reason). Thus one ends up reimplementing the RFS parser, just to figure where to put the format arguments.

In short, this approach is somewhat ignorant towards the Rust compiler, which already dose syntax parsing, and instead employs some "we do it all on our own".


## `macro_rules` Advantages

The big difference when using `macro_rules!` _(or procedural macros which concentrate on tokens instead of string literals only)_ is that one does not have to implement a parser, only specifying a parsable syntax. Thus, it is also easier to change and to extend, and the user has more freedom on how to arrange the tokens, since white spaces are generally ignore with regards to syntax tokens.

So, I claim that my proposal in more flexibly and allows to implement more readable syntax for expressing even complex formattings (as user), while being comparably much easier to implement in Mortal, due to the use of `macro_rules!` and utilizing the parsing provided by the Rust compiler.

Due to this flexibility there are actually lots of concrete syntaxes which can be given to such a macro. And as I mentioned, I suppose it an open question of how to design that concrete syntax and personally I don't have strong preferences. It might be even feasable to implement multiple macros that have different syntaxes. However, I wouldn't recommend to let them differ to much as it would cause confusion among users.

## Summary

Yet as a compromise and to show off the flexibility, I translated your proposal to a more tokenized version compatible with my idea:

```rust
term_writeln!(terminal, <red>("Red")<reset> (" ") <green><bold>("Green")<reset> (" ")
    <green>("{}", 42)<reset>);
term_writeln!(terminal,
    < fg=(Color::Red) >< bg=(Color::Black) > ("Foo")
    <
      reset
    >);
term_writeln!(terminal,<style=(Style::BOLD)>("Bar")<reset>);
```

Notice that the formatting tokens are no longer stringified but actual Rust tokens. Thus, white spaces between strings literals and formatting tokens are of course ignored by the Rust compiler i.e. optional.

This hopefully gives an idea of what I meant by flexibility, also the user has obvieously the option to layout his/her code more flexible, allowing to write complex formatted text more readable.

In conclusion, I strongly oppose putting "all you can" into a single string literal. Instead, I strongly recommend to make use of the tokenization of the Rust compiler and devise a `macro_rules!`-compatible macro syntax, not only because of the sake of easier implementation but also because in the long run it might also benefits the user (as a side note the Rust compiler can give a better error message, once a user messed up his/her macro use).

**Comment by murarth** [comment]
You make some valid points.

The procedural macro solution I presented is perhaps not ideal. I feel it is better, in some ways, than your original proposed solution, but you are correct that it does have its own drawbacks.

Your updated solution does feel like the best parts of each option. Angle brackets and parentheses are unobtrusive and clearly indicate a distinction between color/style elements and formatted text elements.

If you have a working implementation and would like to submit a PR, I would be open to accepting it.

**Comment by Cryptjar** [comment]
I started to implement the macro. First I noticed that while `Terminal`, `TerminalWriteGuard`, `Screen`, and `ScreenWriteGuard` do not share any trait, they all implement the same basic functions needed for the marco, thus one can write a macro which would take any of them as argument. However, there is a different in the return value, the `Terminal*` versions return a `Result` while the `Screen*` versions do not return anything. Therefore, one just ignores the return values (which would cause lots of warings), or implements two different macros (which might cause twice the amount of code). Any Ideas about that?

Next, I reconsidered the syntax: The angular brackets (`<`, `>`) have the big drawback, that they are strictly speaking no _brackets_ to the Rust compiler, but just operators (except in the context of types). Thus, they need to be guarded against other tokens such as expressions.

So, I first tried what works without any additional tokens (such as brackets and commas). When only considering predefined tokens such literal `red` for the foreground color red and literal strings which are not supposed to be Rust Format Strings, one can get an syntax free of brackets and comas. Additionally, other literals such as integers could be simply allowed too as "string literals" and unknown formatting tokens, such as `fancy`here, could be interpreted as `Theme` variable references:

```rust
let fancy = Theme::fg(Color::Green).style(Style::BOLD);
term_writeln!(term red "Red" reset " " fancy "Green" reset " " blue 42);
```

From this point forward, one can add the Rust Format Strings fenced in brackets (any of `()`, `[]`, or `{}`):

```rust
let fancy = Theme::fg(Color::Green).style(Style::BOLD);
let text = "Red";
term_writeln!(term red ("{}", text) reset " " fancy "Green" reset " " blue 42);
```

Next, for the inline fg/bg specification, one can also add an syntax involving brackets. Then, it would be a question of taste whether to allow the predefined tokens (`red`, `blue`, `reset`) without brackets or not. For instance:

```rust
let b : bool;
let fancy = Theme::fg(Color::Green).style(Style::BOLD);
let text = "Red";
term_writeln!(term [fg=Some(Color::Red)] ("{}", text) [reset] " "
    [fancy] "Green" [reset] " "
    [blue] [bg=if b {Some(Color::Yellow)} else {Some(Color::Magenta)}] 42);
```

Alternatively, one can fallback to the angular bracket, however, that would require additional brackets (e.g. `{}`) where ever the user is allowed to place arbitrary expressions:

```rust
let fancy = Theme::fg(Color::Green).style(Style::BOLD);
let text = "Red";
term_writeln!(term <fg={Some(Color::Red)}> ("{}", text) <reset> " "
    <fancy> "Green" <reset> " "
    <blue> <bg={if b {Some(Color::Yellow)} else {Some(Color::Magenta)}}> 42);
```

It would get easier and cleaner the user would be only allowed to specify predefined tokens or a `Theme` variable, because then the user can't specify expressions which would need to be fenced:

```rust
let fancy = Theme::fg(Color::Green).style(Style::BOLD);
let text = "Red";
term_writeln!(term <fg=red> ("{}", text) <reset> " "
    <fancy> "Green" <reset> " " <blue> <bg=yellow> 42);
```

But that obviously denies conditional formatting, except when using `Theme`s.

I found it especially nice, when not having any brackets for the simplest cases while allowing to reference variables for conditional formatting. Additionally, I must admit that in cases where brackets for formattings are required, I found the rectangular brackets (`[]`) prettier that the angular brackets (`<>`) especially because they wound another kind of brackets anyway.

Thus I propose yet another syntax for the implementation:

Color: `red` | `green` | `blue` | ...
Style: `bold` | `italic` | `reverse` | `underline`
Primtive: Color | `#` Color | Style
SimpleFormat: Primtive | `!` Primtive | & Variable |  `reset`
Complex
